datafusion/expr/src/type_coercion/binary.rs

```diff
@@ -287,8 +287,8 @@ fn get_wider_decimal_type(
         (DataType::Decimal128(p1, s1), DataType::Decimal128(p2, s2)) => {
             // max(s1, s2) + max(p1-s1, p2-s2), max(s1, s2)
             let s = *s1.max(s2);
-            let range = (p1 - s1).max(p2 - s2);
-            Some(create_decimal_type(range + s, s))
+            let range = (*p1 as i8 - s1).max(*p2 as i8 - s2);
```

Comment:
Perhaps @liukun4515 you might be able to cast your eyes over this? My knowledge of decimals is fairly limited

Reply:
thanks for the mention for me.
I think arrow ecosystem support negative scale, but in the datafusion we can support `scale>=0`.
We can add checker in the SQL level to ensure `scale >=0` in the datafusion.
cc @alamb

Reply:
In the datafusion, we just support the decimal128, and should make sure the `precision>=1 and precision<=38` and `scale>=0 and scale<=precision`.

Do you have any thoughts for that? @tustvold

Reply:
> but in the datafusion we can support scale>=0

From an outsiders perspective, at least w.r.t decimals, I think breaking compatibility with the arrow specification in this way would likely be surprising to people. Is there a particular reason to not support negative scale?

Reply:
Using the negative scale in the arrow ecosystem is ok for me.
But in the datafusion(SQL level system), it's better to make consistent with other SQL system, like Spark,PG,MySQL.
In other SQL level system, I have not seen the usage of negative scale.

In the PG
```
postgres=# create table test(c1 decimal(10,-1));
ERROR:  NUMERIC scale -1 must be between 0 and precision 10
LINE 1: create table test(c1 decimal(10,-1));
```

In the Spark:

```
spark-sql> create table test_d(c1 decimal(10,-1));
Error in query:
extraneous input '-' expecting INTEGER_VALUE(line 1, pos 34)

== SQL ==
create table test_d(c1 decimal(10,-1))
```

Reply:
Spark definitely supports negative scales, it might not expose them in its SQL frontend though